jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,318 @@
+    {
         @Override
-        public void cancel()
+        public void request(long n)
         {
-            try (AutoLock ignored = lock.lock())
-            {
-                cancelled = true;
-            }
+            // As per rules 3.6 and 3.7, after the Subscription is cancelled all operations MUST be NOPs.
         }
 
-        private void process()
+        @Override
+        public void cancel(LastWill lastWill)
         {
-            while (true)
-            {
-                try (AutoLock ignored = lock.lock())
-                {
-                    if (demand > 0)
-                    {
-                        --demand;
-                    }
-                    else
-                    {
-                        stalled = true;
-                        return;
-                    }
-                }
+            // As per rules 3.6 and 3.7, after the Subscription is cancelled all operations MUST be NOPs.
+        }
+    }
 
-                Content.Chunk chunk = content.read();
+    private static final class ActiveSubscription extends IteratingCallback implements LastWillSubscription
+    {
+        private static final long NO_MORE_DEMAND = -1;
+        private final AtomicReference<LastWill> cancelled;
+        private final AtomicLong demand;
+        private Content.Source content;
+        private Flow.Subscriber<? super Content.Chunk> subscriber;
 
-                if (chunk == null)
+        public ActiveSubscription(Content.Source content, Flow.Subscriber<? super Content.Chunk> subscriber)
+        {
+            this.cancelled = new AtomicReference<>(null);
+            this.demand = new AtomicLong(0);
+            this.content = content;
+            this.subscriber = subscriber;
+        }
+
+        // As per rule 3.3, Subscription MUST place an upper bound on possible synchronous
+        // recursion between Publisher and Subscriber
+        //
+        // As per rule 1.3, onSubscribe, onNext, onError and onComplete signaled to a
+        // Subscriber MUST be signaled serially.
+        //
+        // IteratingCallback guarantee that process() method will be executed by one thread only.
+        // The process() method can be only initiated from request() or cancel() demands methods.
+        @Override
+        protected Action process()
+        {
+            LastWill cancelled = this.cancelled.get();
+            if (cancelled != null)
+            {
+                Throwable reason = cancelled.reason;
+                FinalSignal finalSignal = cancelled.finalSignal;
+                // As per rule 3.13, Subscription.cancel() MUST request the Publisher to eventually
+                // drop any references to the corresponding subscriber.
+                this.demand.set(NO_MORE_DEMAND);
+                // TODO: HttpChannelState does not satisfy the contract of Content.Source "If read() has returned a last chunk, this is a no operation."
```

Comment:
HttpChannelState does not satisfy the contract of Content.Source "If read() has returned a last chunk, this is a no operation."

Reply:
https://github.com/jetty/jetty.project/issues/11879